### DISCUSSION TITLE
ERROR when parsing newline in external scanner.

### DISCUSSION THREAD

**Discussion by m-kru** [tree-sitter/tree-sitter#1281] [Q&A] [answered]
I am trying to prototype some language where newline character `\n` has semantic meaning. However, after moving newline parsing to the external scanner my simple tests cases fail. They always start with `ERROR`  and I have no idea what is wrong. I have been looking at the debug graphs, but I find them rather enigmatic.

My `grammar.js` file looks like this:
```javascript
module.exports = grammar({
  name: 'fbdl',

  extras: $ => [
    $.comment,
  ],

  externals: $ => [
    $.indent,
    $.dedent,
    $.newline
  ],

  rules: {
    source_file: $ => repeat(choice($.meta_statement)),

    identifier: $ => /([A-Z]|[a-z])([A-Z]|[a-z]|[0-9]|_)*/,

    declared_identifier: $ => $.identifier,

    meta_statement: $ => choice(
      $.import_statement,
    ),

    import_statement: $ => choice(
      $.single_import_statement,
    ),

    single_import_statement: $ => seq('import', $.declared_identifier, $.newline)

    comment: $ => token(seq('#', /.*/)),
  }
});
```

and my `scanner.c` file is as follows:
```C
#include <err.h>

#include "tree_sitter/parser.h"

enum TokenType {
	INDENT,
	DEDENT,
	NEWLINE
};

struct Scanner {
	unsigned current_indent;
};

void * tree_sitter_fbdl_external_scanner_create() {
	struct Scanner* scanner = malloc(sizeof(struct Scanner));
	scanner->current_indent = 0;

	if (scanner == NULL)
		err(-1, "Can't allocate memory for external scanner");

	return scanner;
}

void tree_sitter_fbdl_external_scanner_destroy(void *payload) {
	free(payload);
}

unsigned tree_sitter_fbdl_external_scanner_serialize(
	void *payload,
	char *buffer
) {
	return 0;
}

void tree_sitter_fbdl_external_scanner_deserialize(
	void *payload,
	const char *buffer,
	unsigned length
) {
	return;
}

bool tree_sitter_fbdl_external_scanner_scan(
	void *payload,
	TSLexer *lexer,
	const bool *valid_symbols
) {
	struct Scanner *scanner = payload;

	if (valid_symbols[NEWLINE]) {
		if (lexer->lookahead == '\n') {
			lexer->advance(lexer, true);
			lexer->mark_end(lexer);

			while (lexer->lookahead == '\n') {
				lexer->advance(lexer, true);
			}

			lexer->result_symbol = NEWLINE;
			return true;
		}

	}

	return false;
}
```

I have two tests and both fail:

```
======
Import
======

import foo

---
```
```
=======
Comment
=======

# This is single line comment.

---
```

Result of the first one is `(ERROR (newline) (UNEXPECTED 'i') (newline))`, and of the second one is `(ERROR (newline) (comment) (newline))`.

I lack ideas why there is always ERROR at the beginning, I would really appreciate if someone could help me debug this.

**Comment by ahlinc** [comment]
@m-kru I've converted the issue to the discussion because it's the better place for such type of questions.

**Reply by m-kru** [reply]
No problem.

**Comment by ahelwer** [comment]
By default, tree-sitter's `extras` are set up to ignore whitespace (including newlines). If you define your own `extras` - as you have, to support comments - but you still want to ignore (some) whitespace, you have to specify this in your `extras` definition. For example, you could write:
```js
extras: $ => [
  /\s/,
  $.comment,
],
```
which would ignore non-newline whitespace. If you want to ignore newlines too, you can use something like `/\s|\r?\n/`. This will somewhat conflict with your desire to handle newlines in your external scanner, but tree-sitter calls into your external scanner immediately when an external token is expected so it shouldn't be a problem: you'll handle newlines in your external scanner when your grammar says they're expected, and otherwise you'll handle them with `extras`.

Separately, you should also handle leading whitespace in your external scanner. Adding a line like `while (lexer->lookahead == ' ') {lexer->advance(lexer, true);}` to the top of your `scan` function should do the trick. A few other notes about your external scanner:
- Newlines on Windows are `\r\n`, so you'll have to handle that
- You're calling `lexer->advance(lexer, true)` to consume your newline character as whitespace, but it isn't whitespace in the sense that tree-sitter means - that it should be ignored. You don't want to ignore that character so should call `lexer->advance(lexer, false)`; see [here](https://github.com/tree-sitter/tree-sitter/discussions/1255) for more info
- You probably want to call `lexer->mark_end(lexer)` after consuming subsequent newline characters after the first
- Consider how you want your external scanner to behave if there are extraneous spaces on empty lines

If you're writing an external scanner it's also a good idea to become very familiar with `tree-sitter parse -d` output.

**Reply by m-kru** [reply]
@ahelwer thanks for your comments. I was aware of some of them, and I do not handle some of them because I simply want very minimal prototype. However, I have applied few changes like, `/\s/` in extras, `advance(lexer, false),` and `lexer->mark_end(lexer)` after consuming subsequent newline characters after the first. However, I still have the same issue.

You have written:

> but tree-sitter calls into your external scanner immediately when an external token is expected

I wonder how tree-sitter detects the `newline` at the beginning if the `source_file` rule is specified as follows: `    source_file: $ => repeat(choice($.meta_statement)),`.

**Reply by ahelwer** [reply]
@m-kru would have to look at the `tree-sitter parse -d` output to be sure, but it might be tree-sitter encountering a `\n` character at the start of the file, with it not being an extra, so it enters error recovery mode and calls your external scanner with all symbols marked valid; see https://github.com/tree-sitter/tree-sitter/issues/1259

**Comment by m-kru** [comment, marked as answer]
The problem was with following line:

```
    single_import_statement: $ => seq('import', $.declared_identifier, $.newline)
````

Missing comma at the end. I have not idea why it has not been caught at some early parser generation stage.

**Reply by m-kru** [reply]
Ok, so `tree-sitter test` recompiles the `scanner.c`, however if the `grammar.js` has been changed one must manually run `tree-sitter generate`.
